### PR TITLE
Mask body of runOps to avoid a heap corruption

### PR DESCRIPTION
**Pull request by cocreature**

### CODE REVIEW

File: core/src/Network/GRPC/LowLevel/Op.hs [hunk 1]
```diff
@@ -208,7 +208,14 @@ runOps :: C.Call
...
 runOps call cq ops =
   let l = length ops in
-    withOpArrayAndCtxts ops $ \(opArray, contexts) -> do
+    -- It is crucial to mask exceptions here. If we don’t do this, we can
+    -- run into the following situation:
+    --
+    -- 1. We allocate an OpContext, e.g., OpRecvMessageContext and the corresponding ByteBuffer.
+    -- 2. We pass the buffer to gRPC in startBatch.
+    -- 3. If we now get an exception we will free the ByteBuffer.
+    -- 4. gRPC can now end up writing to the freed ByteBuffer and we get a heap corruption.
```
[the remark this answers is not in the document]
Comment (cocreature) [reply core/src/Network/GRPC/LowLevel/Op.hs]: `startBatch` is asynchronous. The actual write happens somewhere in the gRPC internals. I haven’t tracked it down to the exact location.

[the remark this answers is not in the document]
Comment (cocreature) [reply core/src/Network/GRPC/LowLevel/Op.hs]: I took a brief look and it looks like at least `call->receiving_buffer` retains a reference to the buffer https://github.com/grpc/grpc/blob/master/src/core/lib/surface/call.cc#L1804

Probably also worth pointing out that this fix is based on an actual issue that we encountered in our tests.

**Comment by neongreen** [comment]
I can confirm that this patch fixes a segfault that we started to get on an internal project after upgrading to grpc-1.22.

**Comment by Gabriel439** [comment]
@cocreature: Thank you!

**Comment by neongreen** [comment]
Update: this branch works but the `master` of gRPC-haskell doesn't 🤔 https://github.com/awakesecurity/gRPC-haskell/pull/90 seems to be the only difference. @cocreature do your tests pass with the current gRPC-haskell master?

**Comment by cocreature** [comment]
@neongreen I’m confused, this branch is identical to the current master now that the PR is merged.

**Comment by neongreen** [comment]
I believe it's not.

```diff
$ git diff cocreature/fix-heap-corruption origin/master

diff --git a/core/src/Network/GRPC/LowLevel/Op.hs b/core/src/Network/GRPC/LowLevel/Op.hs
index ed8cd8e..bde3997 100644
--- a/core/src/Network/GRPC/LowLevel/Op.hs
+++ b/core/src/Network/GRPC/LowLevel/Op.hs
@@ -108,8 +108,8 @@ freeOpContext (OpSendInitialMetadataContext m _) = C.metadataFree m
 freeOpContext (OpSendMessageContext (bb, s)) =
   C.grpcByteBufferDestroy bb >> C.freeSlice s
 freeOpContext OpSendCloseFromClientContext = return ()
-freeOpContext (OpSendStatusFromServerContext metadata _ _ _) =
-  C.metadataFree metadata
+freeOpContext (OpSendStatusFromServerContext metadata _ _ s) =
+  C.metadataFree metadata >> C.freeSlice s
 freeOpContext (OpRecvInitialMetadataContext metadata) =
   C.metadataArrayDestroy metadata
 freeOpContext (OpRecvMessageContext pbb) =
```

**Comment by neongreen** [comment]
Latest commits in this branch:

```
Mask body of runOps to avoid a heap corruption
Fix sdist tarballs (#88) …
Upgrade to gRPC 1.22 (#85)
```

Latest commits in master:

```
Mask body of runOps to avoid a heap corruption (#89)
Free slice in OpSendStatusFromServerContext (#90) …
Fix sdist tarballs (#88) …
Upgrade to gRPC 1.22 (#85)
```

**Comment by neongreen** [comment]
I'm starting to suspect that this branch is a red herring and the only reason it "fixed" our segfault is that our segfault is caused by #90. I will see if I can share the code.

**Pull request merged**